### PR TITLE
TrieTree implementation

### PR DESCRIPTION
**Pull request by code-monad**
This PR brings a new implementation for `SparseMerkleTree` (check [trie_tree.rs](https://github.com/nervosnetwork/sparse-merkle-tree/pull/34/files#diff-c952d606e11d41c892effa40b10c909e04cb627047052849e920b7aba0b8ece6)), typically called *TrieTree*, which is:

* Less read/write operations while updating compared to the original one
* Less storage space compared to the original one
*  Compatible with the original tree hashes
  - Same root if tree has same leaf
  - Same `MerkleProof`&`Compiled MerkleProof`
  - \*The branches_map won't be the same since the storage is not the same

[Reference](https://ouvrard-pierre-alain.medium.com/sparse-merkle-tree-86e6e2fc26da)

# Read/Write performace comparison:
Here's some bench data with rw operations counters
This can reproduced by the benchmark : [benches/store_counter_benchmark.rs](https://github.com/code-monad/sparse-merkle-tree/blob/master/benches/store_counter_benchmark.rs)
with command: `cargo bench --features trie -- --test` and `cargo bench -- --test`
### New
```
cargo bench --features trie -- --test
#...
random update 100 keys, store counters: Counters { get_branch_counter: 1740, get_leaf_counter: 0, insert_branch_counter: 660, insert_leaf_counter: 100, remove_branch_counter: 0, remove_leaf_counter: 0 }
random update 10000 keys, store counters: Counters { get_branch_counter: 370570, get_leaf_counter: 0, insert_branch_counter: 131628, insert_leaf_counter: 10000, remove_branch_counter: 0, remove_leaf_counter: 0 }
random update_all 100 keys, store counters: Counters { get_branch_counter: 1731, get_leaf_counter: 0, insert_branch_counter: 657, insert_leaf_counter: 100, remove_branch_counter: 0, remove_leaf_counter: 0 }
random update_all 10000 keys, store counters: Counters { get_branch_counter: 371504, get_leaf_counter: 0, insert_branch_counter: 131988, insert_leaf_counter: 10000, remove_branch_counter: 0, remove_leaf_counter: 0 }
```

### Original 
```
cargo bench -- --test
#...
random update 100 keys, store counters: Counters { get_branch_counter: 25600, get_leaf_counter: 0, insert_branch_counter: 25600, insert_leaf_counter: 100, remove_branch_counter: 0, remove_leaf_counter: 0 }
random update 10000 keys, store counters: Counters { get_branch_counter: 2560000, get_leaf_counter: 0, insert_branch_counter: 2560000, insert_leaf_counter: 10000, remove_branch_counter: 0, remove_leaf_counter: 0 }
random update_all 100 keys, store counters: Counters { get_branch_counter: 24836, get_leaf_counter: 0, insert_branch_counter: 24935, insert_leaf_counter: 100, remove_branch_counter: 0, remove_leaf_counter: 0 }
random update_all 10000 keys, store counters: Counters { get_branch_counter: 2418267, get_leaf_counter: 0, insert_branch_counter: 2428266, insert_leaf_counter: 10000, remove_branch_counter: 0, remove_leaf_counter: 0 }
```

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -73,12 +73,18 @@ pub mod merkle_proof;
...
+#[cfg(feature = "trie")]
+pub mod trie_tree;
 
 #[cfg(feature = "smtc")]
 pub use ckb_smt::{SMTBuilder, SMT};
 pub use h256::H256;
 pub use merkle_proof::{CompiledMerkleProof, MerkleProof};
+#[cfg(not(feature = "trie"))]
 pub use tree::SparseMerkleTree;
+pub use tree::{BranchKey, BranchNode};
```
Comment (TheWaWaR) [inline src/lib.rs:85]: Please move out `BranchKey` and `BranchNode` from `tree.rs` to something `branch.rs` and add feature attribute to `pub mod tree` as well.

Comment (quake) [reply src/lib.rs:85]: suggest to keep tree mod private:
https://github.com/code-monad/sparse-merkle-tree/pull/12

Comment (code-monad) [reply src/lib.rs:85]: > suggest to keep tree mod private: [code-monad#12](https://github.com/code-monad/sparse-merkle-tree/pull/12)

This PR has been merged.
IMO, keep `BranchKey`/`BranchNode` in `tree` may refer to a more stable & quick migration

Comment (TheWaWaR) [reply src/lib.rs:85]: > > suggest to keep tree mod private: [code-monad#12](https://github.com/code-monad/sparse-merkle-tree/pull/12)
> 
> This PR has been merged. IMO, keep `BranchKey`/`BranchNode` in `tree` may refer to a more stable & simplified migration

Agreed

File: src/merge.rs [hunk 2]
```diff
@@ -12,6 +12,12 @@ pub enum MergeValue {
         zero_bits: H256,
         zero_count: u8,
     },
+    #[cfg(feature = "trie")]
+    ShortCut {
+        key: H256,
+        value: H256,
+        height: u8,
+    },
```
Comment (TheWaWaR) [inline src/merge.rs:20]: How about add some comment to this enum variant, such as something like:
> A shortcut is the highest subtree root that contains only one key should store the key-value pair instead of its children (the shortcut is still the value of the subtree root that contains 1 non-default value at height 0). With this storage rule, we don’t have to store branches in a sparse tree which significantly reduces the quantity of data to be stored.

File: src/merge.rs [hunk 3]
```diff
@@ -45,6 +71,70 @@ impl MergeValue {
...
+    /// Transform it into a MergeWithZero node
+    #[cfg(feature = "trie")]
+    pub fn into_merge_with_zero<H: Hasher + Default>(&self) -> MergeValue {
+        match self {
+            MergeValue::ShortCut { key, value, height } => {
+                let base_key = key.parent_path(0);
+                let base_node = hash_base_node::<H>(0, &base_key, value);
+                let mut zero_bits = *key;
+                for i in *height..=core::u8::MAX {
+                    if key.is_right(i) {
```
Comment (TheWaWaR) [inline src/merge.rs:125]:
```suggestion
                    if key.get_bit(i) {
```
Use `get_bit` would be more semantically.

File: src/merge.rs [hunk 4]
```diff
@@ -45,6 +71,70 @@ impl MergeValue {
...
+                    }
+                }
+                MergeValue::MergeWithZero {
+                    base_node,
+                    zero_bits,
+                    zero_count: *height,
+                }
+            }
+            _ => {
+                unreachable!();
```
Comment (TheWaWaR) [inline src/merge.rs:136]: Use panic with more meaningful message.

File: src/merge.rs [hunk 5]
```diff
@@ -45,6 +71,70 @@ impl MergeValue {
...
+                zero_bits: _,
+                zero_count: _,
+            } => *base_node,
+            MergeValue::Value(value) => *value,
+        }
+    }
+
+    /// Helper function for Shortcut node
+    /// Transform it into a MergeWithZero node
+    #[cfg(feature = "trie")]
```
Comment (TheWaWaR) [inline src/merge.rs:117]: Since this is a public function, should mention when this will panic.

https://doc.rust-lang.org/src/core/iter/traits/iterator.rs.html#234-237

File: src/merge.rs [hunk 6]
```diff
@@ -45,6 +71,70 @@ impl MergeValue {
...
+                } else {
+                    self.into_merge_with_zero::<H>().hash::<H>()
+                }
+            }
+        }
+    }
+
+    /// Helper function for get base_node
+    /// When call with MergeValue::Value(v), it would be v
+    pub fn base_node<H: Hasher + Default>(&self) -> H256 {
```
Comment (TheWaWaR) [inline src/merge.rs:95]: Where is this function called, or what's the use case?

File: src/merge.rs [hunk 7]
```diff
@@ -123,5 +213,23 @@ fn merge_with_zero<H: Hasher + Default>(
                 zero_count: zero_count.wrapping_add(1),
             }
         }
+        #[cfg(feature = "trie")]
+        MergeValue::ShortCut {
+            key,
+            value,
+            height: _,
```
Comment (TheWaWaR) [inline src/merge.rs:220]:
```suggestion
            ..
```
Otherwise may confused with `height` value in the function parameter.

File: src/merge.rs (hunk 4 above)
Comment (code-monad) [reply src/merge.rs:136]: https://doc.rust-lang.org/std/macro.unreachable.html#panics

According to the doc, we can treat `unreachable` as `panic`, just adding useful messages,while this statement would never reach

File: src/merge.rs (hunk 6 above)
Comment (code-monad) [reply src/merge.rs:95]: This function is not used anymore, it was used to compare the compability of ShortCut nodes and MergedWithZero nodes's equivalence when developing, and can be removed right now

File: src/merge.rs (hunk 4 above)
Comment (TheWaWaR) [reply src/merge.rs:136]: If unreachable only can be sure in current package then this function should not be public.

File: src/trie_tree.rs [hunk 8]
```diff
@@ -0,0 +1,475 @@
...
+                    stack_top -= 1;
+                    skip_height.push(height);
+                }
+                if stack_top == 0 {
+                    break;
+                }
+            }
+
+            let mut proof_result = Vec::new();
+            for height in (0..=fork_height).rev() {
```
Comment (quake) [inline src/trie_tree.rs:367]: suggest to use the filtered heights directly to reduce the `vec#contains` check

```suggestion
             let heights = (0..=fork_height).into_iter().filter(|height|
                if stack_top > 0 && stack_fork_height[stack_top - 1] == *height {
                    stack_top -= 1;
                    false
                } else {
                    true
                }
            ).collect::<Vec<_>>();

            let mut proof_result = Vec::new();
            for height in heights.iter().copied().rev() {
```

File: src/trie_tree.rs [hunk 9]
```diff
@@ -0,0 +1,454 @@
...
+                                (
+                                    MergeValue::shortcut(key, node.hash::<H>(), last_height),
+                                    MergeValue::shortcut(this_key, value, last_height),
+                                )
+                            } else {
+                                (
+                                    MergeValue::from_h256(node.hash::<H>()),
+                                    MergeValue::from_h256(value),
+                                )
+                            };
```
Comment (TheWaWaR) [inline src/trie_tree.rs:156]:
```suggestion
                            } else {
                                if last_height != 0 {
                                    (
                                        MergeValue::shortcut(key, node.hash::<H>(), last_height),
                                        MergeValue::shortcut(this_key, value, last_height),
                                    )
                                } else {
                                    (
                                        MergeValue::from_h256(node.hash::<H>()),
                                        MergeValue::from_h256(value),
                                    )
                                }
                            };
```

File: src/trie_tree.rs [hunk 10]
```diff
@@ -0,0 +1,456 @@
...
+    /// Update multiple leaves at once
+    pub fn update_all(&mut self, mut leaves: Vec<(H256, V)>) -> Result<&H256> {
+        // Dedup(only keep the last of each key) and sort leaves
+        leaves.reverse();
+        leaves.sort_by_key(|(a, _)| *a);
+        leaves.dedup_by_key(|(a, _)| *a);
+
+        for (key, value) in leaves {
+            self.update(key, value)?;
+        }
```
Comment (TheWaWaR) [inline src/trie_tree.rs:282]: @quake 
Do we still need an optimized `update_all()` function?

File: src/trie_tree.rs [hunk 11]
```diff
@@ -0,0 +1,456 @@
...
+
+                match target {
+                    MergeValue::ShortCut {
+                        key: this_key,
+                        value,
+                        height: h,
+                    } => {
+                        if this_key.eq(&key) {
+                            // we update its value
+                            let target = MergeValue::shortcut(key, node.hash::<H>(), h);
```
Comment (TheWaWaR) [inline src/trie_tree.rs:110]:
```suggestion
                            let target = MergeValue::shortcut(key, value.to_h256(), h);
```
Would be more readable.

File: src/trie_tree.rs [hunk 12]
```diff
@@ -0,0 +1,456 @@
...
+                                BranchNode {
+                                    left: next_left,
+                                    right: next_right,
+                                },
+                            )?;
+                            break; // go next round
+                        }
+                    }
+                    _ => {
+                        if target.is_zero() || last_height == 0 {
```
Comment (TheWaWaR) [inline src/trie_tree.rs:174]: It seems in the new shortcut design target.is_zero() is unreachable?

File: src/trie_tree.rs (hunk 10 above)
Comment (quake) [reply src/trie_tree.rs:282]: I don't think it's necessary, because with this PR algorithm, the optimization by `update_all` is very limited (around 1%)

File: src/trie_tree.rs [hunk 13]
```diff
@@ -0,0 +1,456 @@
...
+            debug_assert!(stack_top < MAX_STACK_SIZE);
+            stack_fork_height[stack_top] = fork_height;
+            stack_top += 1;
+            leaf_index += 1;
+        }
+        Ok(MerkleProof::new(leaves_bitmap, proof))
+    }
+}
+
+/// Helper function for a merkle_path insertion
```
Comment (TheWaWaR) [inline src/trie_tree.rs:447]: Rust doc comment is for public function, use `//` instead.

**Review by quake — APPROVED** [review]

File: src/trie_tree.rs (hunk 13 above)
Comment (sopium) [reply src/trie_tree.rs]: FWIW, there's `document-private-items`: https://doc.rust-lang.org/cargo/commands/cargo-doc.html#documentation-options

**Review by TheWaWaR — APPROVED** [review]

**Review by TheWaWaR — CHANGES_REQUESTED** [review]
Add test case for [zero-value shortcut issue](https://github.com/nervosnetwork/sparse-merkle-tree/pull/34/commits/2ab6dee93c95893112e738c4e8ec4f4ebaebb2a4)

**Review by TheWaWaR — APPROVED** [review]
